components/push/src/subscriber.rs

```diff
@@ -175,6 +188,108 @@ impl PushManager {
+}
+
+impl PersistedRateLimiter {
+    pub fn new(op_name: &str, periodic_interval: u64, max_requests_in_interval: u16) -> Self {
+        Self {
+            op_name: op_name.to_owned(),
+            periodic_interval,
+            max_requests_in_interval,
+        }
+    }
+
+    pub fn check(&mut self, store: &Store) -> bool {
+        let (timestamp, count) = self.get_counters(store);
+
+        let now = Self::now_secs();
+        if (now - timestamp) >= self.periodic_interval {
+            log::info!(
+                "Resetting. now({}) - {} < {} for {}.",
+                now,
+                timestamp,
+                self.periodic_interval,
+                &self.op_name
+            );
+            let timestamp = Self::now_secs();
+            self.persist_counters(store, timestamp, 0);
+        } else {
+            log::info!(
+                "No need to reset inner timestamp and count for {}.",
+                &self.op_name
+            )
+        }
+
+        // within interval counter
+        if count > self.max_requests_in_interval {
+            log::info!(
+                "Not allowed: innerCount($innerCount) > $MAX_REQUEST_IN_INTERVAL for {}.",
+                &self.op_name
+            );
+            return false;
+        }
+
+        log::info!("Allowed to pass through for {}!", &self.op_name);
+        // Increment the counter and persist it.
+        self.persist_counters(store, timestamp, count + 1);
+
+        true
+    }
+
+    fn db_meta_keys(&self) -> (String, String) {
+        (
+            format!("ratelimit_{}_timestamp", &self.op_name),
+            format!("ratelimit_{}_count", &self.op_name),
+        )
+    }
+
+    fn get_counters(&self, store: &Store) -> (u64, u16) {
+        (
+            store
+                .get_meta(&format!("ratelimit_{}_timestamp", self.op_name))
+                .ok()
+                .flatten()
+                .map(|s| s.parse())
+                .transpose()
+                .ok()
+                .flatten()
+                .unwrap_or_default(),
+            store
+                .get_meta(&format!("ratelimit_{}_count", self.op_name))
+                .ok()
+                .flatten()
+                .map(|s| s.parse())
+                .transpose()
+                .ok()
+                .flatten()
+                .unwrap_or_default(),
+        )
+    }
+
+    fn persist_counters(&self, store: &Store, timestamp: u64, count: u16) {
+        let (timestamp_key, count_key) = self.db_meta_keys();
+        let r1 = store.set_meta(&timestamp_key, &timestamp.to_string());
+        let r2 = store.set_meta(&count_key, &count.to_string());
+        if r1.is_err() || r2.is_err() {
+            log::warn!("Error updating persisted counters for {}.", &self.op_name);
+        }
+    }
+
+    fn now_secs() -> u64 {
+        SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Current date before unix epoch.")
+            .as_secs()
+    }
+}
```

Comment:
Since we're not re-using `RateLimiter` here, this `PersistedRateLimiter` will need its own suite of tests.